src/video.cpp

```diff
@@ -5,861 +5,908 @@
- * @param pt Base coordinates of the sprite data.
- * @param spr The sprite to blit.
- * @param numx Number of sprites to draw in horizontal direction.
- * @param numy Number of sprites to draw in vertical direction.
- * @param recolour Sprite recolouring definition.
- * @param shift Gradient shift.
+ * Load an OpenGL shader.
+ * @param name Name of the shader.
+ * @return Handle of the loaded shader.
+ */
+GLuint VideoSystem::LoadShader(const std::string &name)
+{
+	std::string v = "data";
+	std::string f = "data";
+	v += DIR_SEP;
+	f += DIR_SEP;
+	v += "shaders";
+	f += "shaders";
+	v += DIR_SEP;
+	f += DIR_SEP;
+	v += name;
+	f += name;
+	v += ".vp";
+	f += ".fp";
+	return this->LoadShaders(FindDataFile(v).c_str(), FindDataFile(f).c_str());
+}
+
+/**
+ * Load an OpenGL shader pair.
+ * @param vp Path to the vertext shader.
+ * @param fp Path to the fragment shader.
+ * @return Handle of the loaded shader pair.
  */
-void VideoSystem::BlitImages(const Point32 &pt, const ImageData *spr, uint16 numx, uint16 numy, const Recolouring &recolour, GradientShift shift)
+GLuint VideoSystem::LoadShaders(const char *vp, const char *fp)
 {
-	this->blit_rect.ValidateAddress();
+	/* Create the shaders. */
+	GLuint vertex_shader_id = glCreateShader(GL_VERTEX_SHADER);
+	GLuint fragment_shader_id = glCreateShader(GL_FRAGMENT_SHADER);
+
+	/* Read the Vertex Shader code from the file. */
+	std::string vertex_shader_code;
+	std::ifstream vertex_shader_stream(vp, std::ios::in);
+	if (vertex_shader_stream.is_open()) {
+		std::stringstream sstr;
+		sstr << vertex_shader_stream.rdbuf();
+		vertex_shader_code = sstr.str();
+		vertex_shader_stream.close();
+	} else {
+		error("ERROR: Unable to open shader '%s'!\n", vp);
+	}
 
-	int x_base = pt.x + spr->xoffset;
-	int y_base = pt.y + spr->yoffset;
+	/* Read the Fragment Shader code from the file. */
+	std::string fragment_shader_code;
+	std::ifstream fragment_shader_stream(fp, std::ios::in);
+	if (fragment_shader_stream.is_open()) {
+		std::stringstream sstr;
+		sstr << fragment_shader_stream.rdbuf();
+		fragment_shader_code = sstr.str();
+		fragment_shader_stream.close();
+	}
 
-	/* Don't draw wildly outside the screen. */
-	while (numx > 0 && x_base + spr->width < 0) {
-		x_base += spr->width; numx--;
+	GLint result = GL_FALSE;
+	int info_log_length;
+
+	/* Compile Vertex Shader. */
+	char const* vertex_source_pointer = vertex_shader_code.c_str();
+	glShaderSource(vertex_shader_id, 1, &vertex_source_pointer, NULL);
+	glCompileShader(vertex_shader_id);
+
+	/* Check Vertex Shader. */
+	glGetShaderiv(vertex_shader_id, GL_COMPILE_STATUS, &result);
+	glGetShaderiv(vertex_shader_id, GL_INFO_LOG_LENGTH, &info_log_length);
+	if (info_log_length > 0) {
+		std::vector<char> vertex_shader_error_message(info_log_length+1);
+		glGetShaderInfoLog(vertex_shader_id, info_log_length, NULL, &vertex_shader_error_message[0]);
+		error("Compile error in shader %s: %s\n", vp, &vertex_shader_error_message[0]);
```

Comment:
Shouldn't you check `result`?  Ie, is it defined that log is empty on successful compile and non-empty on failed compile for all OpenGL systems that exist?

Reply:
I don't see how you addressed this.

Non-empty log does not imply failed compile, see https://stackoverflow.com/questions/56923918/gl-info-log-length-returns-1#56924187